iree/vm/shims.h

```diff
@@ -408,6 +417,7 @@ IREE_VM_ABI_DECLARE_SHIM(r, ii);
 IREE_VM_ABI_DECLARE_SHIM(r, iii);
 IREE_VM_ABI_DECLARE_SHIM(r, iiii);
 IREE_VM_ABI_DECLARE_SHIM(r, r);
+IREE_VM_ABI_DECLARE_SHIM(r, rrrrrr);
```

Comment:
There should be  an analogous call to `IREE_VM_ABI_DEFINE_SHIM` in `shims.c` I guess. @benvanik will know better :)

Reply:
Yep, seems so. Adding Ben as a reviewer therefore.

Reply:
Hmm I'm sketched out by this - why did you need this? There's no HAL calls that return `rrrrrr` and that's making me worry: does emitc need shims for any arbitrary function in a user program? If so, that's going to be a major issue and we'll need to fix that.

Reply:
Currently yes, here is a call stack for a different function.

```c
vm_global_load_i32(uint8_t * base, uint32_t byte_offset) (/home/admin-scamphau/repos/iree/iree/vm/ops.h:21)
module_simple_mul(iree_vm_stack_t * v1, module_t * v2, module_state_t * v3, iree_vm_ref_t * v4, iree_vm_ref_t * v5, iree_vm_ref_t * v6) (/home/admin-scamphau/repos/iree/build/ASAN-EmitC/iree/samples/static_library/simple_mul_emitc.h:347)
call_0rr_r_shim(iree_vm_stack_t * restrict stack, const iree_vm_function_call_t * restrict call, call_0rr_r_t target_fn, void * restrict module, void * restrict module_state, iree_vm_execution_result_t * restrict out_result) (/home/admin-scamphau/repos/iree/iree/vm/shims_emitc.h:303)
iree_vm_native_module_begin_call(void * self, iree_vm_stack_t * stack, const iree_vm_function_call_t * call, iree_vm_execution_result_t * out_result) (/home/admin-scamphau/repos/iree/iree/vm/native_module.c:304)
iree_vm_invoke_within(iree_vm_context_t * context, iree_vm_stack_t * stack, iree_vm_function_t function, const iree_vm_invocation_policy_t * policy, iree_vm_list_t * inputs, iree_vm_list_t * outputs) (/home/admin-scamphau/repos/iree/iree/vm/invocation.c:189)
iree_vm_invoke(iree_vm_context_t * context, iree_vm_function_t function, iree_vm_invocation_flags_t flags, const iree_vm_invocation_policy_t * policy, iree_vm_list_t * inputs, iree_vm_list_t * outputs, iree_allocator_t allocator) (/home/admin-scamphau/repos/iree/iree/vm/invocation.c:218)
iree_runtime_session_call(iree_runtime_session_t * session, const iree_vm_function_t * function, iree_vm_list_t * input_list, iree_vm_list_t * output_list) (/home/admin-scamphau/repos/iree/iree/runtime/session.c:261)
iree_runtime_call_invoke(iree_runtime_call_t * call, iree_runtime_call_flags_t flags) (/home/admin-scamphau/repos/iree/iree/runtime/call.c:96)
Run() (/home/admin-scamphau/repos/iree/iree/samples/static_library/static_library_demo.c:163)
main() (/home/admin-scamphau/repos/iree/iree/samples/static_library/static_library_demo.c:212)
```

Reply:
I just filed https://github.com/google/iree/issues/7842 with more missing shims. Indeed the list keeping increasing could be an issue.

Reply:
Yeah we have two related issues here.

###### Missing shims
There are simply missing shims for calling into the HAL on the EmitC side [here](https://github.com/google/iree/blob/ce1af180c87ed4b4501d4a7b6f2b2617624440c0/iree/vm/shims_emitc.h#L292-L313). There is a 1:1 mapping to the declarations in [shims.h](https://github.com/google/iree/blob/ce1af180c87ed4b4501d4a7b6f2b2617624440c0/iree/vm/shims.h#L405). (On the EmitC side we have to distinguish between variadic and non-variadic calls).

###### Calls to functions in general
Calls to `vm.func` (intra-module) are directly translated to calls to the C functions.

Calls to `vm.import` (inter-module) go through `EMITC_FIXED_IMPORT/EMITC_VLA_IMPORT` and therefore rely on the existing shims for the struct definitions and the auto generated `reset` function.

User calls go through `EMITC_FIXED_SHIM` and also rely on the existing shims for struct definitions and `checked_deref` functions.

###### Possible solution
We could generate all shims in the VM to EmitC conversion, i.e. generate calls to the shim macros instead of hardcoding them in `shims_emitc.h`. We could then build the struct bodies through another variadic macro, and stitch it all together, so the `r -> rrrrrr` call would conceptually contribute something like the following at the beginning of the generated header:

```c
IREE_VM_ABI_FIXED_STRUCT(r, IREE_VM_FIXED_STRUCT_BODY(r))
IREE_VM_ABI_FIXED_STRUCT(rrrrrr, IREE_VM_FIXED_STRUCT_BODY(rrrrrr))
EMITC_DEFINE_SHIMS((r), (r, r, r, r, r, r))
```

Not sure if I'm overlooking something fundamental here.

Reply:
Intra-module calls should not use shims at all - the calling convention stuff is only needed when crossing module boundaries. Like you note shims for inter-module calls should be generated - as there's an infinite number of permutations of those (as it's user code calling across that boundary).

The bytecode gets around needing shims by using the reflected calling convention strings and dynamically packing structs on cross-module calls: https://github.com/google/iree/blob/89e9530b79e7946e04ccde9b0ac226a3f98dbffa/iree/vm/module.c#L37 - which in the emitc case is what you can generate instead. 

Ideally the public functions in an emitc module would take `const iree_vm_function_call_t * call` as args and cast out the structs:
```c
static iree_status_t some_generated_function(iree_vm_stack_t* stack,
                                       const iree_vm_function_call_t* call,
                                       call_i32_i32_t target_fn, void* module,
                                       void* module_state,
                                       iree_vm_execution_result_t* out_result) {
  typedef struct {
    int32_t arg0;
  } args_t;
  typedef struct {
    int32_t ret0;
  } results_t;

  const args_t* args = (const args_t*)call->arguments.data;
  results_t* results = (results_t*)call->results.data;

  ... = args->arg0;
  results->ret0 = ...;
}
```

This hides the implementation detail of the fixed structs and gives the compiler a much better chance of shrinking the code as when bouncing through a shim the compiler cannot inline/propagate across the shim and into the callees - with the above it all becomes just a single pointer lookup instruction into the args/results structs per argument, no extra stack cost, etc. Leads to much smaller and faster code 🎉 

All the shims and such are useful for is humans not wanting to write that boilerplate above but aren't even strictly required for native_module - you could set the pointers yourself to a function like the above. The only reason we have the shims at all is to make it easier to hand-author functions in C - especially when wanting to validate things like variable-length-array lengths and such that would otherwise be easy for a human to accidentally omit. Compilers are nice robots that don't have that problem :)

Reply:
Thanks for the detailed reply! We're on it. Should I open an issue (based on your comment) to keep track?